src/Workspaces/Remote/ServiceHub/Services/DesignerAttribute/RemoteDesignerAttributeIncrementalAnalyzer.cs

```diff
@@ -0,0 +1,181 @@
+            // in this project.
+            var projectVersion = await project.GetDependentSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
+
+            var latestInfos = await ComputeLatestInfosAsync(
+                project, projectVersion, specificDoc, cancellationToken).ConfigureAwait(false);
+
+            // Now get all the values that actually changed and notify VS about them. We don't need
+            // to tell it about the ones that didn't change since that will have no effect on the
+            // user experience.
+            //
+            //  !  is safe here as `i.changed` implies `i.info` is non-null.
+            var changedInfos = latestInfos.Where(i => i.changed).Select(i => i.info!.Value).ToList();
+            if (changedInfos.Count > 0)
+            {
+                await _endPoint.InvokeAsync(
+                    nameof(IDesignerAttributeServiceCallback.RegisterDesignerAttributesAsync),
+                    new object[] { changedInfos },
+                    cancellationToken).ConfigureAwait(false);
+            }
+
+            // now that we've notified VS, persist all the infos we have (changed or otherwise) back
+            // to disk.  We want to do this even when the data is unchanged so that our version
+            // stamps will be correct for the next time we come around to analyze this project.
+            //
+            // Note: we have a potential race condition here.  Specifically, for simplicity, the VS
+            // side will return immediately, without actually notifying the project system.  That
+            // means that we could persist the data to local storage that isn't in sync with what
+            // the project system knows about.  i.e. if VS is closed or crashes before that
+            // information is persisted, then these two systems will be in disagreement.  this is
+            // believed to not be a big issue given how small a time window this would be and how
+            // easy it would be to get out of that state (just edit the file).
+
+            await PersistLatestInfosAsync(projectVersion, latestInfos, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task PersistLatestInfosAsync(VersionStamp projectVersion, (Document, DesignerInfo? info, bool changed)[] latestInfos, CancellationToken cancellationToken)
+        {
+            foreach (var (doc, info, _) in latestInfos)
+            {
+                // Skip documents that didn't change contents/version at all.  No point in writing
+                // back out the exact same data as before.
+                if (info == null)
+                    continue;
+
+                using var memoryStream = new MemoryStream();
+                using var writer = new ObjectWriter(memoryStream);
+
+                info.Value.WriteTo(writer, projectVersion);
+
+                memoryStream.Position = 0;
+                await _storage.WriteStreamAsync(
```

Comment:
Can this throw IOException?

Reply:
couldn't find anything in the docs for the apis we call indicating that can happen.  and i don't believe we've seen that.  this currently matches the rest of our surface area for using the persistence service.  if IOException was possible, i think we would move the fix *into* the persistence service to handle it and return `false` (to indicate it failed).